python/src/aiconfig/default_parsers/openai_vision.py

```diff
@@ -0,0 +1,427 @@
+        with vision does not support the message.name parameter, functions/tools,
+        response_format parameter
+
+        Args:
+            prompt (str): The prompt to be serialized.
+            inference_settings (dict): Model-specific inference settings to be serialized.
+
+        Returns:
+            str: Serialized representation of the prompt and inference settings.
+        """
+        event = CallbackEvent(
+            "on_serialize_start",
+            __name__,
+            {
+                "prompt_name": prompt_name,
+                "data": data,
+                "parameters": parameters,
+                "kwargs": kwargs,
+            },
+        )
+        await ai_config.callback_manager.run_callbacks(event)
+        prompts: list[Prompt] = []
+
+        # Combine conversation data with any extra keyword args
+        conversation_data = {**data}
+
+        if not "messages" in conversation_data:
+            raise ValueError(
+                "Data must have `messages` array to match openai api spec"
+            )
+
+        # Find first system prompt. Every prompt in the config will bet set to use this system prompt.
+        system_prompt = None
+        for message in conversation_data["messages"]:
+            if "role" in message and message["role"] == "system":
+                # Note: system prompt for openai represented will be an object with content and role attributes {content: str, role: str}
+                system_prompt = message
+                break
+
+        # Get the global settings for the model
+        model_name = (
+            conversation_data["model"]
+            if "model" in conversation_data
+            else self.id()
+        )
+
+        model_metadata = ai_config.get_model_metadata(
+            conversation_data, model_name
+        )
+        # Remove messages array from model metadata. Handled separately
+        if model_metadata.settings:
+            model_metadata.settings.pop("messages", None)
+
+        # Add in system prompt
+        if system_prompt:
+            if not model_metadata.settings:
+                model_metadata.settings = {}
+            model_metadata.settings["system_prompt"] = system_prompt
+
+        i = 0
+        while i < len(conversation_data["messages"]):
+            message = conversation_data["messages"][i]
+            role = message["role"]
+            if role == "user":
+                # Serialize User message as a prompt and save the assistant response as an output
+                assistant_response: Union[ChatCompletionMessage, None] = None
+                if i + 1 < len(conversation_data["messages"]):
+                    next_message = conversation_data["messages"][i + 1]
+                    if next_message["role"] == "assistant":
+                        assistant_response = next_message
+                        i += 1
+                new_prompt_name = f"{prompt_name}_{len(prompts) + 1}"
```

Comment:
Same as comment below, why not use `i` since we can now have prompts with the same name

Reply:
This is the exact same logic from existing openai parser, minus the function and tools message handling (since that's not supported for vision).

using len(prompts) won't have the same name since prompts is added to with user message here. I don't know if `i` would work as well since `i` is incremented for each message (not just user message) so you could end up with prompt_name and then prompt_name_3 if there are assistant messages between user messages

Reply:
Ah I see, thanks this makes sense